### PR TITLE
Freeze fuse two mms

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111233
* __->__ #111232

Improves llama_v2 perf locally from 1.48x -> 1.55x. 

A good future rewrite would be to unify the freezing batching with the other batching rules that @yanboliang & co were working on. I want to wait for the forthcoming pre-dispatch changes to settle down first.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by Chillee** [comment]
batching rules you say??

**Comment by eellison** [comment]
To be more clear, this batches horizontal mms that share a single input where the other inputs are parameters, and constant folds the concat of the params in freezing.

File: test/inductor/test_inductor_freezing.py [hunk 1]
```diff
@@ -174,6 +174,16 @@ def __init__(self):
             def forward(self, x):
                 return x @ self.t1, x @ self.t2, x @ self.t3
 
+        class MM2(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+                self.t1 = torch.nn.Parameter(torch.rand(10, 10))
+                self.t2 = torch.nn.Parameter(torch.rand(10, 10))
```
Comment (jansel) [inline test/inductor/test_inductor_freezing.py:182]: Can you add a test where the sizes of t1/t2 aren't the same?

File: torch/_inductor/fx_passes/freezing_patterns.py [hunk 2]
```diff
@@ -148,6 +150,24 @@ def matmul_replacement(inp, w1, w2, w3):
         exclusive_arg_names=("w1", "w2", "w3"),
     )
 
+    def matmul_fuse_pattern_two(inp, w1, w2):
+        return (inp @ w1, inp @ w2)
+
+    def matmul_replacement_two(inp, w1, w2):
+        cat_t = torch.cat((w1, w2), dim=1)
+        mm = inp @ cat_t
+        return mm.chunk(2, dim=1)
```
Comment (jansel) [inline torch/_inductor/fx_passes/freezing_patterns.py:159]: I think this assumes the sizes are equal.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/freezing_patterns.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/fx_passes/freezing_patterns.py:159]: That condition is checked above:

https://github.com/pytorch/pytorch/blob/bfccf138b02bc4093d5990abdaac71c487842805/torch/_inductor/fx_passes/freezing_patterns.py#L130-L133

Comment (jansel) [reply torch/_inductor/fx_passes/freezing_patterns.py:159]: Ah, it should be possible to handle the mismatched case too with a slice.

Comment (eellison) [reply torch/_inductor/fx_passes/freezing_patterns.py:159]: will leave that for a follow up

Comment (jansel) [reply torch/_inductor/fx_passes/freezing_patterns.py:159]: Sounds good, do add the test though.

**Review by Chillee — APPROVED** [review]

**Pull request closed without merging**